utils/zerovec/src/ule/niche.rs

```diff
@@ -0,0 +1,173 @@
+    pub fn new(opt: Option<U>) -> Self {
+        assert!(N == core::mem::size_of::<U>());
+        match opt {
+            Some(u) => Self { valid: u },
+            None => Self {
+                niche: <U as NicheBytes<N>>::NICHE_BIT_PATTERN,
+            },
+        }
+    }
+
+    /// Convert to an Option<U>
+    pub fn get(self) -> Option<U> {
+        // Safety: The union stores NICHE_BIT_PATTERN when None otherwise a valid U
+        unsafe {
+            if self.niche == <U as NicheBytes<N>>::NICHE_BIT_PATTERN {
+                None
+            } else {
+                Some(self.valid)
+            }
+        }
+    }
+}
+
+impl<U: NicheBytes<N> + ULE, const N: usize> Copy for NichedOptionULE<U, N> {}
+
+impl<U: NicheBytes<N> + ULE, const N: usize> Clone for NichedOptionULE<U, N> {
+    fn clone(&self) -> Self {
+        *self
+    }
+}
+
+impl<U: NicheBytes<N> + ULE + PartialEq, const N: usize> PartialEq for NichedOptionULE<U, N> {
+    fn eq(&self, other: &Self) -> bool {
+        self.get().eq(&other.get())
+    }
+}
+
+impl<U: NicheBytes<N> + ULE + Eq, const N: usize> Eq for NichedOptionULE<U, N> {}
+
+/// Safety for ULE trait
+/// 1. NichedOptionULE does not have any padding bytes due to `#[repr(C)]` on a struct
+///    containing only ULE fields.
+///    NichedOptionULE either contains NICHE_BIT_PATTERN or valid U byte sequences.
+///    In both cases the data is initialized.
+/// 2. NichedOptionULE is aligned to 1 byte due to `#[repr(packed)]` on a struct containing only
+///    ULE fields.
+/// 3. validate_byte_slice impl returns an error if invalid bytes are encountered.
+/// 4. validate_byte_slice impl returns an error there are extra bytes.
+/// 5. The other ULE methods are left to their default impl.
+/// 6. NichedOptionULE equality is based on ULE equality of the subfield, assuming that NicheBytes
+///    has been implemented correctly (this is a correctness but not a safety guarantee).
+unsafe impl<U: NicheBytes<N> + ULE, const N: usize> ULE for NichedOptionULE<U, N> {
+    fn validate_byte_slice(bytes: &[u8]) -> Result<(), crate::ZeroVecError> {
+        let size = size_of::<Self>();
+        // The implemention is only correct if NICHE_BIT_PATTERN has same number of bytes as the
+        // type.
+        debug_assert!(N == core::mem::size_of::<U>());
+
+        // The bytes should fully transmute to a collection of Self
+        if bytes.len() % size != 0 {
+            return Err(crate::ZeroVecError::length::<Self>(bytes.len()));
+        }
+        bytes.chunks(size).try_for_each(|chunk| {
+            // Associated const cannot be referenced in a pattern
+            // https://doc.rust-lang.org/error-index.html#E0158
+            if chunk == <U as NicheBytes<N>>::NICHE_BIT_PATTERN {
+                Ok(())
+            } else {
+                U::validate_byte_slice(chunk)
+            }
+        })
+    }
+}
+
+/// Optional type which uses [`NichedOptionULE<U,N>`] as ULE type.
+/// Derefs to [`Option<U>`]
```

Comment:
Issue: Where did the Deref impl go? Please either add it back or remove this comment

Reply:
Sorry I missed that comment. Removed.